### PR TITLE
os: Re-adding registry-authentication.md

### PR DESCRIPTION
**Pull request by ElijahCaine**
People in the community want this document back.

I'm not sure what's wrong with this, I know I removed it in the first place but it seems fine. I admit it might be a bit lacking in breadth (it only talks about Docker registry auth, but it could talk about rkt, k8s, and mezos auth (reading off of what quay.io provides 😜 ).

Thoughts?

### CODE REVIEW

**Comment by ElijahCaine** [comment]
I'll ask the Quay folks what they think this doc should include, but for now I started writing the registry-auth doc I would expect to see. Edits and feedback welcome.

**Note** There's an un-edited screengrab with a secret in this doc. I did delete the bot associated with that screenshot so I'm 99% sure that's okay. I kept it clean because it's useful information for users to show the reader.

File: os/registry-authentication.md [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+
+![Quay Robot settings][quay-bot-img]
+
+[comment]: # (Yes, this robot has been deleted. Un-obscured to help maintain educational value.)
+
+For more information, check out the [Quay robot documentation][quay-robot].
+
+## Manual registry auth setup
+
+### Docker hub
```
Comment (joshix) [inline os/registry-authentication.md]: Docker Hub is a registry; the others in this list are properly not. They may belong in this document, but not at the identical level as a registry *to which* one authenticates.

File: os/registry-authentication.md [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+### Kubernetes
+
+### rkt
+
+### Mezos
+
+[quay-site]: https://quay.io/
+[docker-hub-site]: https://hub.docker.com/
+[quay-robot]: https://docs.quay.io/glossary/robot-accounts.html
```
Comment (joshix) [inline os/registry-authentication.md]: A verification of the age and currency of this document is called for

**Comment by joshix** [comment]
ref #914 #917

**Comment by joshix** [comment]
@ElijahCaine What's wrong with the original (removed) reg-auth.md is that it is obsolete; this much is recorded in #914. I just don't remember the details of why it was ok to leave it unreplaced (and should have recorded same in the PR.) 

@josephschorr Did the deletion in #914 (eventually #917) arise from a conversation between you and I in Raleigh?

**Comment by ElijahCaine** [comment]
@joshix Added some tweaks. Will continue writing the doc throughout today.

Regarding the relevancy of that quay doc link: I asked some of the team quay folks and they basically said it's out of date but there isn't anything better on the quay side of things. If it becomes an issue I'd be glad to overhaul that doc too. Apparently their quay docs are pretty stagnant compared with the quay-enterprise docs, but those don't cover a super-set of topics.

**Comment by ElijahCaine** [comment]
@robszumski @joshix I'd like to get this merged back in soon so we can get back to people that asked about their 404's. Any feedback on the completed first draft would be appreciated.

File: os/registry-authentication.md [hunk 3]
```diff
@@ -0,0 +1,199 @@
+# Using authentication for a registry
+
+To pull images behind an authentication wall you need to setup some form of authentication with your image registry. These are either [quay.io][quay-site] or [hub.docker.com][docker-hub-site] or a self-hosted `registry.example.io/v0`.
+
+This document outlines the preferred method of authentication with an image registry like [quay.io][quay-site] and [hub.docker.com][docker-hub-site] and how setup registry auth for Docker, Kubernetes, rkt, and Mezos.
```
Comment (robszumski) [inline os/registry-authentication.md]: Mesos

File: os/registry-authentication.md [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+
+## Manual registry auth setup
+
+If you are using a registry other than Quay (e.g., Docker Hub, Docker Store, etc) you will need to manually configure your credentials with your container-runtime or orchestration tool.
+
+### Docker
+
+*Quay robots provide Docker registry authentication config files and intructions on to set them up. This can be found by accessing a robot's settings in the* **Robot Accounts** *tab under your user settings*.
+
+#### $HOME/.docker/config.json
```
Comment (robszumski) [inline os/registry-authentication.md]: Can we make this describe the section better? "Writing the config file"? Do we even need this heading?

File: os/registry-authentication.md [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+            "xxxx": "xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx"
+        },
+        "https://registry.example.io/v0/": {
+            "auth": "xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx="
+        }
+    }
+}
+```
+
+This `config.json` file can be copied to `/home/core/.docker/config.json` on a CoreOS node during first-boot using [ignition][ignition]. If this file is copied the `core` users's `$HOME` directory, ensure that `core` is the user executing `docker` commands otherwise the credentials will not be used.
```
Comment (robszumski) [inline os/registry-authentication.md]: Can we give the full ignition example?

File: os/registry-authentication.md [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+
+Docker also offers the ability to configure a credentials store. This is outlined in-depth on the [docker login documentation page][docker-login].
```
Comment (robszumski) [inline os/registry-authentication.md]: What does "configure a credentials store" mean?

File: os/registry-authentication.md [hunk 7]
```diff
@@ -0,0 +1,199 @@
...
+
+Docker is adding more fine-grained permission control (similar to Quay robots) in [Docker Registry v2][docker-reg-v2], but as of early 2017 that is not yet fully integrated into the Docker CLI or Docker Hub interface.
+
+### Kubernetes
+
+Kubernetes uses secrets to store registry credentials.
+
+*As mentioned above, Quay robots provide this Kubernetes secret and instructions on how to use it in the* **Robot Accounts** *tab under your user settings*.
```
Comment (robszumski) [inline os/registry-authentication.md]: These caveats scattered throughout make this hard to follow/read

File: os/registry-authentication.md [hunk 8]
```diff
@@ -0,0 +1,199 @@
...
+  
+  imagePullSecrets:
+    - name: my-favorite-registry-secret
+```
+
+For more information read the [docker-registry Kubernetes secret][k8s-docker-registry] and [Kubernetes imagePullSecrets][k8s-image-pull] documents on kubernetes.io.
+
+### rkt
+
+Rkt store registry-authentication in a JSON file stored in the directory `/etc/rkt/auth.d/`.
```
Comment (robszumski) [inline os/registry-authentication.md]: `rkt` (lowercase)

Comment (robszumski) [inline os/registry-authentication.md]: `stores`

File: os/registry-authentication.md [hunk 9]
```diff
@@ -0,0 +1,199 @@
...
+## Using a Quay robot for registry auth
+
+The preferred way to setup registry authentication with [quay.io][quay-site] is to use a [Quay Robot][quay-robot]. This essentially acts as a login-token wth some nice features including:
+
+- Pre-populated repository-auth configuration files.
+- Credentials are limited to specific repositories.
+- Choose from read, write, or admin privileges.
+- Token regeneration.
+
+![Quay Robot settings][quay-bot-img]
```
Comment (robszumski) [inline os/registry-authentication.md]: It'd probably be a good idea to crop this image so it doesn't contain your username in it

File: os/registry-authentication.md (hunk 3 above)
Comment (robszumski) [inline os/registry-authentication.md]: Should we add a table of contents up here?

Comment (ElijahCaine) [inline os/registry-authentication.md]: @robszumski coreos.com docs have a table of contents on the sidebar. I can add another (manual) TOC if you think it'd be necessary. I can see pros for it. The main con is that it'd have to be hard-coded for it to show up on GitHub.

Comment (robszumski) [inline os/registry-authentication.md]: Good point, I think that one is fine

File: os/registry-authentication.md [hunk 10]
```diff
@@ -0,0 +1,234 @@
...
+      "file:///path/to/registry.example.io.tar.gz"
+  ]
+}
+```
+
+More thorough information about configuring Mesos registry authentication can be found on the ['Using a Private Docker Registry'][mesos-registry] Mesos/Marathon documentation.
+
+## Copying the config file with ignition
+
+[Ignition][ignition] is a low level system configuration utility used to provision a CoreOS node on it's first boot. Here is an example of copying one of the configuration files mentioned above onto the appropriate path in the node:
```
Comment (robszumski) [inline os/registry-authentication.md]: I feel like we're missing what this gets you. "Doing this on first boot allows you machine to immediately have access to all your private images blah blah"

**Comment by joshix** [comment]
@ElijahCaine Merge at will. This will need topmatter again in the -pages repo.
